ddtrace/contrib/openai/patch.py

```diff
@@ -0,0 +1,540 @@
+        except StopIteration:
+            pass
+    finally:
+        span.finish()
+        integration.metric(span, "dist", "request.duration", span.duration_ns)
+
+
+def _patched_endpoint(integration, patch_hook):
+    @trace_utils.with_traced_module
+    def patched_endpoint(openai, pin, func, instance, args, kwargs):
+        g = _traced_endpoint(patch_hook, integration, openai, pin, instance, args, kwargs)
+        g.send(None)
+        resp, err = None, None
+        try:
+            resp = func(*args, **kwargs)
+            return resp
+        except Exception as e:
+            err = e
+            raise
+        finally:
+            try:
+                g.send((resp, err))
+            except StopIteration:
+                pass
+
+    return patched_endpoint
+
+
+def _patched_endpoint_async(integration, patch_hook):
+    # Same as _patched_endpoint but async
+    @trace_utils_async.with_traced_module
+    async def patched_endpoint(openai, pin, func, instance, args, kwargs):
+        g = _traced_endpoint(patch_hook, integration, openai, pin, instance, args, kwargs)
+        g.send(None)
+        resp, err = None, None
+        try:
+            resp = await func(*args, **kwargs)
+            return resp
+        except Exception as e:
+            err = e
+            raise
+        finally:
+            try:
+                g.send((resp, err))
+            except StopIteration:
+                pass
+
+    return patched_endpoint
+
+
+class _EndpointHook:
+    def handle_request(self, integration, span, args, kwargs):
+        raise NotImplementedError
+
+
+class _BaseCompletionHook(_EndpointHook):
+    """Completion and ChatCompletion share a lot of logic, capture that here."""
+
+    _request_tag_attrs = []
+
+    def _record_request(self, span, kwargs):
+        for kw_attr in self._request_tag_attrs:
+            if kw_attr in kwargs:
+                if isinstance(kwargs[kw_attr], dict):
+                    set_flattened_tags(
+                        span, [("request.{}.{}".format(kw_attr, k), v) for k, v in kwargs[kw_attr].items()]
+                    )
+                else:
+                    span.set_tag("request.%s" % kw_attr, kwargs[kw_attr])
+
+
+class _CompletionHook(_BaseCompletionHook):
+    _request_tag_attrs = [
+        "suffix",
+        "max_tokens",
+        "temperature",
+        "top_p",
+        "n",
+        "stream",
+        "logprobs",
+        "echo",
+        "stop",
+        "presence_penalty",
+        "frequency_penalty",
+        "best_of",
+        "logit_bias",
+        "user",
+    ]
+
+    def handle_request(self, integration, span, args, kwargs):
+        sample_pc_span = integration.is_pc_sampled_span(span)
+
+        if sample_pc_span:
+            prompt = kwargs.get("prompt", "")
+            if isinstance(prompt, list):
+                for idx, p in enumerate(prompt):
+                    span.set_tag_str("request.prompt.%d" % idx, integration.trunc(p))
```

Comment:
Should we add a tag for the prompt length? and even set `1` when we don't have a list?

(random thought, feel free to resolve)

Reply:
Consider, what data would be useful when we don't have `sample_pc_span`

Reply:
Out of scope for now, thanks for the suggestion though - improving prompt/completion representations beyond the literal text has been marked down as future work